zcash_primitives/src/merkle_tree.rs

```diff
@@ -44,590 +23,307 @@ pub trait HashSer {
-/// tree.append(cmu);
-/// witness.append(cmu);
-/// assert_eq!(tree.root(), witness.root());
-/// ```
-#[derive(Clone)]
-pub struct IncrementalWitness<Node: Hashable> {
-    tree: CommitmentTree<Node>,
-    filled: Vec<Node>,
-    cursor_depth: usize,
-    cursor: Option<CommitmentTree<Node>>,
+pub fn write_nonempty_frontier_v1<H: HashSer, W: Write>(
+    mut writer: W,
+    frontier: &NonEmptyFrontier<H>,
+) -> io::Result<()> {
+    write_position(&mut writer, frontier.position())?;
+    if frontier.position().is_odd() {
+        // The v1 serialization wrote the sibling of a right-hand leaf as an optional value, rather
+        // than as part of the ommers vector.
+        frontier
+            .ommers()
+            .get(0)
+            .expect("ommers vector cannot be empty for right-hand nodes")
+            .write(&mut writer)?;
+        Optional::write(&mut writer, Some(frontier.leaf()), |w, n: &H| n.write(w))?;
+        Vector::write(&mut writer, &frontier.ommers()[1..], |w, e| e.write(w))?;
+    } else {
+        frontier.leaf().write(&mut writer)?;
+        Optional::write(&mut writer, None, |w, n: &H| n.write(w))?;
+        Vector::write(&mut writer, frontier.ommers(), |w, e| e.write(w))?;
+    }
+
+    Ok(())
 }
 
-impl<Node: Hashable> IncrementalWitness<Node> {
-    /// Creates an `IncrementalWitness` for the most recent commitment added to the given
-    /// [`CommitmentTree`].
-    pub fn from_tree(tree: &CommitmentTree<Node>) -> IncrementalWitness<Node> {
-        IncrementalWitness {
-            tree: tree.clone(),
-            filled: vec![],
-            cursor_depth: 0,
-            cursor: None,
-        }
-    }
-
-    /// Reads an `IncrementalWitness` from its serialized form.
-    #[allow(clippy::redundant_closure)]
-    pub fn read<R: Read>(mut reader: R) -> io::Result<Self> {
-        let tree = CommitmentTree::read(&mut reader)?;
-        let filled = Vector::read(&mut reader, |r| Node::read(r))?;
-        let cursor = Optional::read(&mut reader, CommitmentTree::read)?;
-
-        let mut witness = IncrementalWitness {
-            tree,
-            filled,
-            cursor_depth: 0,
-            cursor,
-        };
-
-        witness.cursor_depth = witness.next_depth();
-
-        Ok(witness)
-    }
-
-    /// Serializes this `IncrementalWitness` as an array of bytes.
-    pub fn write<W: Write>(&self, mut writer: W) -> io::Result<()> {
-        self.tree.write(&mut writer)?;
-        Vector::write(&mut writer, &self.filled, |w, n| n.write(w))?;
-        Optional::write(&mut writer, self.cursor.as_ref(), |w, t| t.write(w))
-    }
+#[allow(clippy::redundant_closure)]
+pub fn read_nonempty_frontier_v1<H: HashSer + Clone, R: Read>(
+    mut reader: R,
+) -> io::Result<NonEmptyFrontier<H>> {
+    let position = read_position(&mut reader)?;
+    let left = H::read(&mut reader)?;
+    let right = Optional::read(&mut reader, H::read)?;
+    let mut ommers = Vector::read(&mut reader, |r| H::read(r))?;
```

Comment:
```suggestion
    let mut ommers = Vector::read(&mut reader, H::read)?;
```
Why not this rather than `#[allow(clippy::redundant_closure)]`?

Reply:
IIRC this didn't work in the past for reasons, but maybe our MSRV is smart enough now?

Reply:
Does not compile.